### PR TITLE
expression: rewrite builtin function: FROM_UNIXTIME

### PR DESCRIPTION
**Pull request by XuHuaiyu**
PTAL @zz-jason @coocood 
to #4080

### CODE REVIEW

**Comment by zz-jason** [comment]
@XuHuaiyu ci failed.

**Comment by XuHuaiyu** [comment]
PTAL @zz-jason

File: expression/builtin_time.go [hunk 1]
```diff
@@ -1154,41 +1174,62 @@ func (b *builtinFromUnixTimeSig) eval(row []types.Datum) (d types.Datum, err err
...
+
+func (b *builtinFromUnixTime1ArgSig) evalTime(row []types.Datum) (res types.Time, isNull bool, err error) {
+	return evalFromUnixTime(b.ctx, b.tp.Decimal, row, b.args[0])
+}
+
+type builtinFromUnixTime2ArgSig struct {
+	baseStringBuiltinFunc
+}
+
+func (b *builtinFromUnixTime2ArgSig) evalString(row []types.Datum) (res string, isNull bool, err error) {
```
Comment (zz-jason) [inline expression/builtin_time.go]: add a comment here

File: expression/builtin_time.go [hunk 2]
```diff
@@ -1154,41 +1174,62 @@ func (b *builtinFromUnixTimeSig) eval(row []types.Datum) (d types.Datum, err err
...
+	return t, false, nil
+}
 
-	if args[0].Kind() == types.KindString { // Keep consistent with MySQL.
-		t.Fsp = types.MaxFsp
+type builtinFromUnixTime1ArgSig struct {
+	baseTimeBuiltinFunc
+}
+
+func (b *builtinFromUnixTime1ArgSig) evalTime(row []types.Datum) (res types.Time, isNull bool, err error) {
```
Comment (zz-jason) [inline expression/builtin_time.go]: ditto

File: expression/builtin_time.go [hunk 3]
```diff
@@ -1154,41 +1174,62 @@ func (b *builtinFromUnixTimeSig) eval(row []types.Datum) (d types.Datum, err err
...
 	if err != nil {
-		return d, errors.Trace(err)
+		return res, true, errors.Trace(err)
 	}
 	nano := new(types.MyDecimal).FromInt(int64(time.Second))
 	x := new(types.MyDecimal)
 	err = types.DecimalMul(fracDecimalTp, nano, x)
 	if err != nil {
-		return d, errors.Trace(err)
+		return res, isNull, errors.Trace(err)
```
Comment (zz-jason) [inline expression/builtin_time.go]: return `res, true, errors.Trace(err)`

File: expression/builtin_time.go [hunk 4]
```diff
@@ -1108,40 +1109,59 @@ type fromUnixTimeFunctionClass struct {
...
 	if unixTimeStamp.IsNegative() {
-		return
+		return res, true, nil
 	}
 	integralPart, err := unixTimeStamp.ToInt()
-	if err == types.ErrTruncated {
+	if terror.ErrorEqual(err, types.ErrTruncated) {
 		err = nil
 	}
 	if err != nil {
```
Comment (zz-jason) [inline expression/builtin_time.go]:
```go
if err != nil && !terror.ErrorEqual(err, types.ErrTruncated) {
    ...
}
```

File: expression/builtin_time.go [hunk 5]
```diff
@@ -1154,41 +1174,62 @@ func (b *builtinFromUnixTimeSig) eval(row []types.Datum) (d types.Datum, err err
...
-		return d, errors.Trace(err)
+		return res, isNull, errors.Trace(err)
 	}
 	fractionalPart, err := x.ToInt() // here fractionalPart is result multiplying the original fractional part by 10^9.
-	if err == types.ErrTruncated {
+	if terror.ErrorEqual(err, types.ErrTruncated) {
 		err = nil
 	}
 	if err != nil {
```
Comment (zz-jason) [inline expression/builtin_time.go]:
```go
if err != nil && !terror.ErrorEqual(err, types.ErrTruncated) {
    ...
}
```

**Comment by zz-jason** [comment]
@XuHuaiyu plz resolve conflicts

**Comment by zz-jason** [comment]
@jackysp @winkyao @breeswish PTAL

**Pull request merged**